### DISCUSSION TITLE
Disabling Search Index Auto Updating on Entry Save

### DISCUSSION THREAD

**Discussion by ben-mcfetridge-resn** [statamic/cms#5316] [Q&A] [answered]
Does Statamic 3 have the option to control the auto indexing of search anymore? I haven't been able to find anything about it.

I have Algolia setup totally fine, my issue is that I don't want Statamic to update the record within Algolia on every save of an entry. I'd rather do it manually with the command php please search:update.

In Statamic 2 there was a setting in site/settings/search.yaml with auto_index: true which would then in statamic/bundles/Search/SearchListener.php either submit the updated change to Algolia (or chosen search engine) or not on save of the page/entry (https://v2.statamic.com/knowledge-base/performance#disable-search-autoindexing). There doesn’t appear to be documentation on this within Statamic 3 that I can find. I assume that there should be a similar function to the insert function within SearchListener.php or whatever the equivalent is within Statamic 3.

Might help to find it within Statamic; in Statamic 2 the function which is doing the search update is (in SearchListener.php):

```
 /**
     * Insert content into the search index
     *
     * @param RepositoryItemInserted $event
     */
    public function insert(RepositoryItemInserted $event)
    {
        if (! Config::get('search.auto_index')) {
            return;
        }
.... (edited)
```

**Comment by edalzell** [comment]
> I have Algolia setup totally fine, my issue is that I don't want Statamic to update the record within Algolia on every save of an entry. I'd rather do it manually with the command php please search:update.

Why?

**Comment by ben-mcfetridge-resn** [comment]
We have migrated to Statamic 3 and are using the headless version of it with Nuxt, which means that we deploy just a front end build rather than with the CMS as it's a static site. We also have a situation with a site where we don't control the deployment/server of the final live site and merely provide the static built site to be deployed.

As a result of this the CMS where edits are done are on our hosted servers. We have a search index for that CMS build in order to test etc but also have a separate search index for the deployed front end. Otherwise the changes that are being made within the CMS and haven't been deployed yet will show up in search results on the live environment (due to all search indices being updated on save) - we obviously don't want this misalignment and don't want to potentially expose non published content within search results.

So ideally we are able to turn off the auto indexing, as Statamic 2 had, and we can control manually using the `php please search:update` when doing a production build to ensure the search index for production matches the content actually deployed.

If there's another/better way to manage this obviously open to that too, with it ideally being as automated as possible.

**Reply by edalzell** [reply]
use a different algolia index for staging/local?

**Reply by ben-mcfetridge-resn** [reply]
I might be misunderstanding your suggestion, sorry if I am. 

We do have different Algolia indexes, in this simplified case, we have one for the CMS and it's direct frontend preview build and one for the live site which is a static build built from the CMS data.

My key issue is how do we:

1. Only update one of those indices when an entry is saved within Statamic? Only the CMS's front end index and not the deployed live site index.
2. Update the search index for the deployed front end **only** when a build is done, effectively creating the search index for the deployed live site as that point in time.

I have tried:

- Adding mutliple indexes within `config/statamic/search.php` but they all get updated on a save of any record.
- Duplicating the Algolia index at build time but we are unfortunately not able to go that route as from [Algolia's documentation](https://www.algolia.com/doc/api-reference/api-methods/copy-index/) "Copying a source index that has replicas copies the index’s data, but not its replicas. You can’t copy to a destination index that already has replicas." In our situation we have replicas.

So to me the ideal situation is that we turn off the auto indexing on save by Statamic and manually control the updating of each of the search indexes as part of the build process.

**Comment by jasonvarga** [comment, marked as answer]
It will always update. But sounds like you have a bit of a custom situation. You can override the event subscriber.

In your AppServiceProvider, add this to `register`.

```php
$this->app->extend(UpdateItemIndexes::class, function ($original) {
    return config('statamic.search.auto_update')
        ? $original
        : new DontUpdateItemIndexes;
});
```

and create the `DontUpdateItemIndexes` class which is basically a dummy override.

```php
<?php

namespace App;

class DontUpdateItemIndexes
{
    public function subscribe($event)
    {
        // do nothing
    }
}
```

Then add the new config key, which you can control however makes sense for you, in this case through an `.env` var.

```php
// config/statamic/search.php

<?php

return [

    'auto_update' => env('SEARCH_AUTO_UPDATE', false),

    // ...

];

**Reply by ben-mcfetridge-resn** [reply]
The dream answer! Thank you so much, I thought because Statamic 2 had the option that it was hidden away somewhere in Statamic 3 that I had missed. Thank you again!

**Comment by moritzlang** [comment]
Thanks for the solution @jasonvarga!
I would also appreciate adding the option again in Statamic 3, since I think my use case is not that uncommon:

The frontend consists of just static html files (SSG). I want the search index to be based on the same content state as when building the statics. Therefore I want to update the search index only when building the frontend.
